flax/linen/module.py

```diff
@@ -0,0 +1,415 @@
+# -----------------------------------------------------------------------------
+def all_names_on_object(x):
+  """Get all names on self and on classes throughout MRO."""
+  nameset = set(x.__dict__.keys())
+  for cls in x.__class__.__mro__:
+    nameset = nameset.union(set(cls.__dict__.keys()))
+  return nameset
+
+# Method wrapping of "compact methods" and setup()
+# -----------------------------------------------------------------------------
+def compact(fun):
+  """Allow inline submodules and parameters within a single Module method."""
+  @functools.wraps(fun)
+  def wrapped_method(self, *args, **kwargs):
+    # Store compact method *name* rather than reference, so that subclassing
+    # a module with a @compact __init__ method is still possible
+    if getattr(self, '_compact_method', fun.__name__) != fun.__name__:
+      raise RuntimeError(
+          'Only one method per class can be @compact. You can remove @compact '
+          'and define submodules and variables in setup(), or use two '
+          'separate modules.')
+    object.__setattr__(self, '_compact_method', fun.__name__)
+
+    if self.scope is None:
+      raise ValueError("Can't call methods on orphaned modules")
+
+    object.__setattr__(self, '_in_compact_method', True)
+    _context.module_stack.append(self)
+    try:
+      return fun(self, *args, **kwargs)
+    finally:
+      _context.module_stack.pop()
+      object.__setattr__(self, '_in_compact_method', False)
+      object.__setattr__(self, '_reservations', set())
+      object.__setattr__(self, '_autoname_cursor', dict())
+      object.__setattr__(self, '_last_varname', None)
+      object.__setattr__(self, 'scope', self.scope.rewound())
+  return wrapped_method
+
+def wrap_setup(fun):
+  @functools.wraps(fun)
+  def wrapped_setup_method(self, *args, **kwargs):
+    object.__setattr__(self, '_in_setup', True)
+    try:
+      return fun(self, *args, **kwargs)
+    finally:
+      object.__setattr__(self, '_in_setup', False)
+  return wrapped_setup_method
+
+
+# Base Module definition.
+# -----------------------------------------------------------------------------
+class Module:
+  """Base Module Class"""
+  # class defaults -- Note: these __must not__ be made annotations!
+  _in_compact_method = False
+  _in_call = False
+  _in_setup = False
+  _last_varname = None
+  scope = None
+
+  @classmethod
+  def __init_subclass__(cls):
+    """Automatically initialize all subclasses as custom dataclasses."""
+    # All Flax Modules are dataclasses.  We force this convention since
+    # it encourages the stateless behavior needed to clone module instances for
+    # functional transformation.  Instead of using a python metaclass, we
+    # automatically transform Modules into dataclasses at subclass creation
+    # time, and we set the first dataclass argument to `parent`, and the last
+    # optional argument to `name`.
+    cls._add_parent_and_name_attrs()
+    dataclasses.dataclass(cls)
```

Comment:
`__eq__` seems safe(?) and is set by default.  Because we define a `__setattr__` we can't set `frozen=True`, we could however mimic the same frozen semantics ourselves, as well as implement a custom `__hash__` if we want to.